### DISCUSSION TITLE
ReadResponseAsString behavior

### DISCUSSION THREAD

**Discussion by John-Papachristos** [RicoSuter/NSwag#4482] [Q&A] [answered]
Hi there, 

I wanted to ask about the intended behavior of the ReadResponseAsString option. I was hoping it would return the API response as a string to the caller, but it looks like the generated code shown below is still returning an object:

                        if (status_ == 200)
                        {
                            var objectResponse_ = await ReadObjectResponseAsync<Message>(response_, headers_, cancellationToken).ConfigureAwait(false);
                            if (objectResponse_.Object == null)
                            {
                                throw new ApiException("Response was null which was not expected.", status_, objectResponse_.Text, headers_, null);
                            }
                            return objectResponse_.Object;

If I change the generated code to return the Text property instead, then it gives me the result I was hoping for:

                   return objectResponse_.Text;

However, I don't want to change the generated code. I was wondering if the generated code is working as intended, if it should actually be returning the Text property when the ReadResponseAsString flag is set to true.

Thanks,
John P.

**Comment by SebastianStehle** [comment]
The goal of this property is not to return a string, but to read it as string. 

The HTTP response body is made available as a string. Therefore you can read directly from that and avoid the allocation of the string. But this has the disadvantage that deserialization errors are harder to debug. Therefore there is a property to configure that. For example you can set the property to true in you test or staging environment and set it to false in production to have better performance.

**Reply by John-Papachristos** [reply]
Thanks for that clarification. I guess I misunderstood the intent of this property. However, it's unfortunate that the string response is not exposed in some way by the NSwag client to the calling process, since I don't want to make any changes to the generated code. 

But underlying my question in general, is the way the generated code returns an instance of a specific type, instead of a generic object. The REST API that I'm importing has a base class called "Message" of type object defined. It also has a number of classes that derive from this base class. All operations in the API will return a response of type Message, and it contains a property that identifies the specific type to use for the response. Then we can cast the response object to the specific type and extract the data from it. 
But the generated code for the NSWag client returns an instance of the base type (Message) which contains only the data that has been deserialized for this type from the response content. Attempting to cast this object to a derived type does not give us the data we need from the response. And this is why I was interested in the ReadResponseAsString property, because it contains the data we need for deserializing to the required derived type. Ideally, there would be a way to return a generic object in this situation. 

We have used AutoRest in the past and surprisingly it behaves as we need it to in this regard, We felt that NSWag may be a better option for us going forward, but it seems there is no workaround for this issue without changing the generated code. Unless there are other options I could try?

Thanks again for taking the time to respond to my question. 

John P.

**Reply by SebastianStehle** [reply]
It is exposed in exception afaik

**Reply by John-Papachristos** [reply]
Yes, I noticed that it was included when an exception is thrown. Unfortunately not so in the OK response case:

return new SwaggerResponse<Message>(status_, headers_, objectResponse_.Object);

Ideally, what I would like to see is the objectResponse_.Text included as part of this SwaggerResponse when ReadResponseAsString is turned on. Something like this:

return new SwaggerResponse<Message>(status_, headers_, objectResponse_.Object, objectResponse_.Text);

Would this be a worthwhile inclusion to the product, to cater for the situation that I explained earlier?

Thanks,
John P.

**Reply by SebastianStehle** [reply]
I am not the Autor but I would not add that because it is easy enough to edit the templates. All these special cases need to be maintained

**Reply by John-Papachristos** [reply]
Thanks. That's great! I wasn't aware that the generation templates were available to customize. It looks like this could be an answer for us. I'll give it a try.

**Reply by SebastianStehle** [reply, marked as answer]
I create everything from a custom console program and have a few template customizations: 
https://github.com/Squidex/squidex-samples/blob/master/csharp/Squidex.ClientLibrary/CodeGeneration/Program.cs#L40

**Reply by John-Papachristos** [reply]
I tried this and it works well. Opens up a whole new world of possibilities. Thank you very much for your help with this.
